### DISCUSSION TITLE
[JPA] 빠르게 QDSL 사용하는 방법

### DISCUSSION THREAD

**Discussion by ahngilwoong** [cs-collections/cs-collections-discussion#56] [Spring]
### QDSL이란?
 
QDSL(Querydsl)은 Java 런타임에서 SQL 또는 JPQL(Java Persistence Query Language) 쿼리를 생성하는데 사용할 수 있는 도구이다.
QDSL을 사용하면, 동적으로 SQL 또는 JPQL 쿼리를 만들 수 있다.

### QDSL 적용하기

우선 Maven 의존성을 추가하고
```maven
<dependency>
    <groupId>com.querydsl</groupId>
    <artifactId>querydsl-jpa</artifactId>
</dependency>
<dependency>
    <groupId>com.querydsl</groupId>
    <artifactId>querydsl-apt</artifactId>
    <exclusions>
        <exclusion>
            <groupId>org.eclipse.jdt.core.compiler</groupId>
            <artifactId>ecj</artifactId>
	    </exclusion>
    </exclusions>
</dependency>
<dependency>
    <groupId>com.querydsl</groupId>
    <artifactId>querydsl-sql</artifactId>
    <version>4.1.4</version>
</dependency>
<dependency>
    <groupId>com.querydsl</groupId>
    <artifactId>querydsl-sql-spring</artifactId>
    <version>4.1.4</version>
</dependency>
```

pom.xml build안의 plugins 안에 이 플러그인을 추가 해준다.
```maven
            <plugin>
                <groupId>com.mysema.maven</groupId>
                <artifactId>apt-maven-plugin</artifactId>
                <version>1.1.3</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>process</goal>
                        </goals>
                        <configuration>
                            <outputDirectory>target/generated-sources/java</outputDirectory>
                            <processor>com.querydsl.apt.jpa.JPAAnnotationProcessor</processor>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
```

그 다음에 QDSL에서 사용할 Qclass를 만들어주려면
프로젝트에서 마우스 오른쪽 클릭 > Maven > Generate Sources and Update Folders 메뉴 클릭
![image](https://user-images.githubusercontent.com/65801290/216225473-14dcc555-c020-4523-b580-be55b045acf5.png)
을 해주면 준비는 끝났고

QClass 파일이 존재함에도 불구하고 실행 시 Cannot resolve symbol 'class명' 오류가 발생하면
이 경우 File > Project Structure 메뉴를 클릭 한 후 generated 폴더를 Sources로 설정한다.

출처 => https://yjh5369.tistory.com/entry/IntelliJ-Maven-%ED%99%98%EA%B2%BD%EC%97%90%EC%84%9C-QueryDsl-%EC%82%AC%EC%9A%A9%ED%95%98%EB%8A%94-%EB%B0%A9%EB%B2%95

그 후 JPA를 사용한 Repository 구조는 
<img width="1175" alt="image" src="https://user-images.githubusercontent.com/65801290/216225656-ab7ce883-bed2-426e-8a94-ab1a59db1d76.png">
이렇게 되고 실제 QDSL이 적용된 클래스는

```java
import com.example.productservice.productstore.jpo.MemberJPO;
import com.example.productservice.productstore.jpo.QMemberJPO;
import com.querydsl.jpa.impl.JPAQuery;
import java.util.List;
import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import lombok.RequiredArgsConstructor;


@RequiredArgsConstructor
public class MemberRepositoryImpl implements MemberRepositoryCustom{
    @PersistenceContext
    private final EntityManager em;

    @Override
    public List<MemberJPO> findMembersAll(){
        QMemberJPO qMemberJPO = QMemberJPO.memberJPO;
        JPAQuery<MemberJPO> query = new JPAQuery<>(em);
        return query
            .select(qMemberJPO)
            .from(qMemberJPO)
            .fetch();
    }
}
```
이런 모습이다.

QDSL(Querydsl)은 JPA, SQL, JPQL 등을 사용한 동적 쿼리 작성을 단순화하는 Java 프레임워크이다. QDSL에서 사용되는 문법은 다음과 같다:

Select clause: select(path) 메소드를 사용하여 조회할 필드를 정의한다.

From clause: from(path) 메소드를 사용하여 쿼리의 조회 대상 테이블을 정의한다.

Where clause: where(predicate) 메소드를 사용하여 쿼리의 조건을 정의한다.

Order by clause: orderBy(orderSpecifier) 메소드를 사용하여 정렬 조건을 정의한다.

Group by clause: groupBy(groupExpression) 메소드를 사용하여 그룹 지정을 정의한다.

Having clause: having(predicate) 메소드를 사용하여 그룹 제약 조건을 정의한다.

QDSL에서 제공하는 편의적인 API를 사용하여 간단한 문법으로 동적 쿼리를 작성할 수 있다. 예를 들어, where() 메소드에서 지정한 조건을 결합할 때 and()와 or() 메소드를 사용할 수 있다.

fetch는 다음과 같다.
fetch() : 조회 결과가 여러 건일 때 사용
fetchOne() : 조회 결과가 1건만 있을 때 사용 (1건 이상이면 NonUniqueResultException 발생)
fetchFirst() : 무조건 첫 번째건 반환
fetchResults() : 조회 결과 + 개수 반환
fetchCount() : 조회 결과의 개수 반환

QDSL의 where절 안에 조건을 위한 메소드들이 있는데 
goe() 쿼리에서 비교를 위한 조건자를 만들기 위한 Querydsl의 메서드이다.
goe() 는 비교보다 크거나 같음을 만드는 데 사용됩니다. 
예를 들어:
```java
QEmployee employee = QEmployee.employee;
List<Employee> result = queryFactory
  .selectFrom(employee)
  .where(employee.salary.goe(50000))
  .fetch();
```
월급이 50000보다 크거나 같은 것을 찾는것이다.

eq() 은 "Equal To"를 나타내며 필드 값이 지정된 값과 같아야 하는 쿼리에서 조건을 지정하는 데 사용됩니다. 
예를 들어:
```java
QEmployee employee = QEmployee.employee;
Employee result = queryFactory
  .selectFrom(employee)
  .where(employee.firstName.eq("John"))
  .fetchOne();
```
직원 이름의 첫글자가 John인 사람을 찾는것이다,

**Comment by Irisation23** [comment]
혹시 해당 기술 카테고리 Spring으로 옮겨도 괜찮을까요?

**Reply by ahngilwoong** [reply]
넵
